Hotfix/certification deploy dummy

rmelotte opened the pull request:
We have an issue in the certification testbed where the dummy prplMesh is not always stopped.

This PR deploys the dummy build to /opt/prplMesh, so that it's always accessible on subsequent runs, to fix the issue.

It also uses `ci/run_easymesh_cert.sh` in Gitlab CI, so that both the nightly tests on the intel testbed and the tests in CI use the same script.

It requires the following change in the easymesh_cert repository: https://git.prpl.dev/prplmesh/wfa-certification/easymesh_cert/-/merge_requests/12

rmelotte committed:
CI: run_easymesh_cert.sh: add a function to deploy the dummy build

This is a preparatory commit.

When running a certification test on the WFA testbed, each instance of
prplMesh is first stopped using the prplmesh_utils.sh script. It's
done before starting any test and not after, to prevent
failed/canceled job from polluting the current one.

Since the new certification test jobs do not fetch the dummy build
when the DUT is the rax40, the prplmesh_utils.sh cannot be used to
stop the dummy prplMesh.

To solve this issue, the following 3 possibilities were considered:

- stop the dummy prplMesh manually without relying on
  prplmesh_utils.sh.

- always pull in the dummy build regardless of what the DUT is.

- when running the test with the dummy prplMesh, deploy it somewhere
  else so that the script that started it is always available when
  testing other devices.

Option 1 will not be robust in the long run, since it will need to be
adapted every time there is a change to the way prplMesh is stopped.

Option 2 has the disadvantage that the version of the script used to
stop prplMesh could be different than the version of the script that
started it.

For those reasons, the option 3 was chosen.

At the moment we call the deploy_ipk.sh script from the .gitlab_ci.yml
configuration, but also from ci/run_easymesh_cert.sh (for nightly
tests on the intel testbed).

To avoid further duplicating the deploy logic, run_easymesh_cert.sh is
adapted to make it possible to deploy the dummy prplMesh, and
.gitlab-ci.yml will later be changed to use run_easymesh_cert.sh.

Making Gitlab CI use "run_easymesh_cert.sh" also makes it easier to
later upload logs to owncloud.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
CI: certification: use run_easymesh_cert.sh

Using run_easymesh_cert.sh allows us to move the deploy logic away
from the CI configuration, to make it possible to upload logs to
owncloud, and also unifies the way the different testbeds are used.

.gitlab-ci.yml:

- switch to ci/run_easymesh_cert.sh. We don't need a condition to
  deploy prplMesh anymore, since it will be handled by the script
  itself.

ci/certification/generic.yml:

- the "echo $CI_COMMIT_DESCRIPTION" was a leftover and should have
been removed, so remove it now.

- switch to ci/run_easymesh_cert.sh.

ci/certification/netgear-rax40.yml:

- remove the before_script, since it's now handled by
  run_easymesh_cert.sh from the generic job.

The ssh device names are the same as the $DEVICE_UNDER_TEST, so we can
use the same value for both the "-s" and the "-d" options of
ci/run_easymesh_cert.sh.

We also need to use the "--skip-download" option because Gitlab CI
takes care of that.

Since we want to make sure basic tests still pass, run them:
MAP-5.3.1:dummy
MAP-4.2.1:netgear-rax40

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
CI: run_easymesh_cert.sh: add an option to skip downloading prplMesh

If we want to use run_easymesh_cert.sh in Gitlab CI, we have to add
the ability to skip the download, as Gitlab already takes care of that
with the dependencies.

- add a new option, --skip-download, to skip the download (but not
  necessarily the upgrade). It defaults to false, to keep the same
  behavior as before.

- move the downloaded ipk to build/$TARGET_DEVICE. Since that's where
  Gitlab CI downloads the ipk, it's easier to put the downloaded one
  there.

- only store the ipk and build info if an ipk was deployed.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
CI: run_easymesh_cert.sh: save the test results from run_test_file

If we want to be able to use run_easymesh_cert.sh in CI, we also need
the script to exit with an code that corresponds to the test results.

Save the exit code from run_test_file.sh, and exit
run_easymesh_cert.sh with it.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
CI: run_easymesh_cert.sh: fix rootdir

owncloud_definitions.sh overrides the rootdir value.

As a quick fix, source it before setting rootdir so that
run_easymesh_cert.sh can be used again.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

arnout commented on ci/run_easymesh_cert.sh:
```diff
@@ -54,29 +56,47 @@ is_prplmesh_device() {
+}
+
 upgrade_prplmesh() {
+    if [ "$TARGET_DEVICE" = prplmesh ] ; then
+        deploy_dummy_prplmesh && return 0 || return 1
+    fi
```
I think it would be nicer to do:
```suggestion
    if [ "$TARGET_DEVICE" = prplmesh ] ; then
        deploy_dummy_prplmesh
    else    
        deploy_prplwrt_prplmesh
    fi
```
Note that the return is implied: the return value of the function is the one of the last statement. So if we have nothing after this, that is sufficient. But maybe a bit fragile.

arnout commented on ci/run_easymesh_cert.sh:
```diff
@@ -54,29 +56,47 @@ is_prplmesh_device() {
 
+deploy_dummy_prplmesh() {
+    info "Deploy prplmesh locally to /opt/prplMesh"
+    rm -rf /opt/prplMesh
```
Is this script run as root?

rmelotte replied:
~~Yes, because run_test_file.sh needs to be run as root.~~
Actually no, this particular script isn't run as root, which means the runner will have to have write permissions on /opt

arnout commented on ci/run_easymesh_cert.sh:
```diff
@@ -54,29 +56,47 @@ is_prplmesh_device() {
+    info "Deploy prplmesh locally to /opt/prplMesh"
+    rm -rf /opt/prplMesh
+    mkdir -p /opt/prplMesh
+    if ! mv "$rootdir/build/install" /opt/prplMesh ; then
```
Err, won't this put the installation under /opt/prplMesh/install instead of under /opt/prplMesh?

More importantly, however: AFAIR, prplMesh isn't relocatable. When you move it somewhere else, some things won't work anymore. For example configuration file paths are hardcoded. So if we want to install it there, we should also build with CMAKE_INSTALL_PREFIX=/opt/prplMesh (and then install with DESTDIR=...). And that is a much bigger change I'm afraid...

rmelotte replied:
Indeed that's annoying...

Unless you see an issue with that, I'll update the CI configuration to always build with the prefix set to `/opt/prplMesh`.

arnout replied:
> Unless you see an issue with that, I'll update the CI configuration to always build with the prefix set to `/opt/prplMesh`.

I see an issue with that! test_flows.py assumes that it's installed in rootdir/build/install... So, if we move it to /opt/prplMesh (which is a good thing IMO), we should also change test_flows.py to start with a deployment that puts it there, and change maptools.py to use DESTDIR during installation. So it's quite a big change...

rmelotte replied:
Can't we do the deployment outside of test_flows.py and update test_flows.py to point to /opt/prplMesh?

rmelotte committed:
tools: download_ipk.sh: honor --path when downloading from owncloud

When using download_ipk.sh with the "--path" option to download the
files from owncloud to a specific path, DOWNLOAD_PATH was not used, so
the downloaded files would always be in the current directory.

Add "$DOWNLOAD_PATH" like we do when downloading from Gitlab.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte commented:
Building and deploying to /opt could be an annoying change for people working and testing on the same shared machine, so closing for now.
